File: TTdsa-cpp/Exercise/array/N01.cpp
//
// Created by renxin on 2021/6/8.
//

#include "ttdsa.h"

class Solution {
public:
    bool Find(int target, vector<vector<int> > array) {
        if (!array.empty()) {
            int r = 0;
            int c = array[0].size() - 1;
            while (r < array.size() && c >= 0) {
                if (array[r][c] < target) {
                    ++r;
                } else if (target < array[r][c]) {
                    --c;
                } else {
                    return true;
                }
            }
        }
        return false;
    }
};

int main() {
    Solution sol;
    vector<vector<int>> array = {{1, 2, 8,  9},
                                 {2, 4, 9,  12},
                                 {4, 7, 10, 13},
                                 {6, 8, 11, 15}};
    // 1
    cout << sol.Find(7, array) << endl;
    // 0
    cout << sol.Find(3, array) << endl;

    return 0;
}

File: TTdsa-java/Exercise/src/com/aacirq/array/N01.java
package com.aacirq.array;

public class N01 {
    public static void main(String[] args) {
        N01 solution = new N01();

        int[][] nums = {{1, 2, 8, 9}, {2, 4, 9, 12}, {4, 7, 10, 13}, {6, 8, 11, 15}};
        // true
        System.out.println(solution.Find(7, nums));
        // false
        System.out.println(solution.Find(3, nums));
    }

    public boolean Find(int target, int[][] array) {
        if (array.length == 0)
            return false;
        int r = 0;
        int c = array[0].length - 1;
        while (r < array.length && c >= 0) {
            if (target < array[r][c]) {
                --c;
            } else if (target > array[r][c]) {
                ++r;
            } else {
                return true;
            }
        }
        return false;
    }
}
